justfile

```diff
@@ -1,63 +1,170 @@
+        {{ if rs-build-type == "release" { "--release" } else { "" } }} \
+        {{ _features }} \
+        {{ flags }}
+
+# Compile Rust unit tests
+rs-test-build:
+    {{ _cargo }} test --no-run --frozen \
+        --workspace \
+        {{ _features }} \
+        {{ _fmt }}
 
-#
-# Recipes
-#
+# Run Rust unit tests
+rs-test *flags:
+    {{ _cargo }} {{ _cargo-test }} --frozen \
+        {{ if rs-build-type == "release" { "--release" } else { "" } }} \
+        {{ _features }} \
+        {{ flags }}
 
-# Run formatting, tests, and build the iptables container
-default: fmt test build
+# Check each crate independently to ensure its Cargo.toml is sufficient.
+rs-check-dirs:
+    #!/usr/bin/env bash
+    set -euo pipefail
+    while IFS= read -r toml ; do
+        {{ just_executable() }} \
+            rs-build-type='{{ rs-build-type }}' \
+            rs-features='{{ rs-features }}' \
+            rs-toolchain='{{ rs-toolchain }}' \
+            _rs-check-dir "${toml%/*}"
+        {{ just_executable() }} \
+            rs-build-type='{{ rs-build-type }}' \
+            rs-features='{{ rs-features }}' \
+            rs-toolchain='{{ rs-toolchain }}' \
+            _rs-check-dir "${toml%/*}" --tests
+    done < <(find . -mindepth 2 -name Cargo.toml | sort -r)
+
+_rs-check-dir dir *flags:
+    cd {{ dir }} \
+        && {{ _cargo }} check --frozen \
+                {{ if rs-build-type == "release" { "--release" } else { "" } }} \
+                {{ _features }} \
+                {{ flags }} \
+                {{ _fmt }}
+
+# If we're running in github actions and cargo-action-fmt is installed, then add
+# a command suffix that formats errors.
+_fmt := if env_var_or_default("GITHUB_ACTIONS", "") != "true" { "" } else {
+    ```
+    if command -v cargo-action-fmt >/dev/null 2>&1; then
+        echo "--message-format=json | cargo-action-fmt"
+    fi
+    ```
+}
+
+# Configures which features to enable when invoking cargo commands.
+_features := if rs-features == "all" {
+        "--all-features"
+    } else if rs-features != "" {
+        "--no-default-features --features=" + rs-features
+    } else { "" }
+
+# Use nextest if it's available (i.e. when running locally).
+_cargo-test := ```
+    if command -v cargo-nextest >/dev/null 2>&1; then echo " nextest run"
+    else echo " test" ; fi
+    ```
+
+##############
+# PROXY-INIT #
+##############
 
 # Build the project
-build:
-    go build -o out/linkerd2-proxy-init main.go
+go-build:
+    go build -o out/linkerd2-proxy-init ./proxy-init
 
 # Runs Go's code formatting tool and succeeds if no output is printed
-fmt:
+go-fmt:
     gofmt -d .
     test -z "$(gofmt -d .)"
 
 # Run unit tests
-test-unit:
+go-test-unit:
     go test -v ./...
 
 # Run integration tests
-test-integration cluster='init-test':
+go-test-integration cluster='init-test': docker-proxy-init docker-tester
     k3d image import -c {{ cluster }} {{ docker_tester_tag }} {{ docker_tag }}
     cd integration_test && ./run_tests.sh
```

Comment:
we should pull in changes from main that decouple image loading and test running -- but I don't think the integration tests should be "go" or "rust", really. instead, they should probably be `proxy-init-test` and `validator-test`, etc.

Some of the top level recipes like `go-fmt` and `rs-fmt` probably do make sense as language scoped commands, since they operate on all sources for a language in the repo; but it's an implementation detail that the proxy-init integration tests are written in Go (and really, they're driven by a bash script and not `go test`)...

Reply:
Sounds good!